Merge Mindstorms EV3 adapters & demo to main repository

oblaufus opened the pull request.

kjsmita6 commented:
This looks really cool and great work! I'll take a look.

kjsmita6 commented on components/mindstorms-ev3/pom.xml:
```diff
@@ -0,0 +1,45 @@
+   
+    <modelVersion>4.0.0</modelVersion>
+
+    <groupId>com.example</groupId>
```
com.tibco.ep.community.components

kjsmita6 commented on components/mindstorms-ev3/pom.xml:
```diff
@@ -0,0 +1,45 @@
+    <description></description>
+    
+    <licenses>
+    <license>
```
Could you fix this indentation? From here to `</licenses>` should be indented.

kjsmita6 commented on components/mindstorms-ev3/src/main/eventflow/com/example/ev3/Demo.sbapp:
```diff
@@ -0,0 +1,225 @@
+        </annotations>
+        <type-metadata>
+            <param name="type" value="module"/>
+            <param name="fully-qualified-name" value="com.example.ev3.Demo"/>
```
Same as above, `com.example` should be `com.tibco.ep.community.components` everywhere.

kjsmita6 commented on components/mindstorms-ev3/src/main/eventflow/com/example/ev3/Demo.sbapp:
```diff
@@ -0,0 +1,225 @@
+<?xml version="1.0" encoding="UTF-8" standalone="no"?>
```
For both `.sbapp` files, can you format it (Cmd+L) so the arrows are drawn nicely?

kjsmita6 commented on components/mindstorms-ev3/src/main/eventflow/com/example/ev3/EV3.sbapp:
```diff
@@ -0,0 +1,160 @@
+                <field name="TargetPort" type="string"/>
+            </schema>
+        </stream>
+        <box name="Command" type="java">
```
Should this be connected to something? It's just floating.

kjsmita6 commented on components/mindstorms-ev3/src/main/eventflow/com/example/ev3/Demo.sbapp:
```diff
@@ -0,0 +1,225 @@
+                <field name="Time" type="int"/>
+            </schema>
+        </stream>
+        <box name="Connection" type="java">
```
Your adapter will need icons eventually.

kjsmita6 commented on components/mindstorms-ev3/src/main/eventflow/com/example/ev3/SensorColors.sbint:
```diff
@@ -0,0 +1,20 @@
+        <param name="fully-qualified-name" value="com.example.ev3.SensorColors"/>
+    </type-metadata>
+    <memory-model-settings/>
+    <constant expr="&quot;BLACK&quot;" name="Black"/>
```
Nice use of constants.

kjsmita6 commented on components/mindstorms-ev3/src/main/java/com/example/ev3/EV3CommandAdapter.java:
```diff
@@ -0,0 +1,390 @@
+
+    // Expected motor command schema
+    private static Schema.Field FIELD_COMMAND = Schema.createField(DataType.STRING, "Command");
+    private boolean StreamPortA;
```
My Eclipse tells me that these `private boolean` fields are not used anywhere.

kjsmita6 commented on components/mindstorms-ev3/src/main/java/com/example/ev3/EV3CommandAdapter.java:
```diff
@@ -0,0 +1,390 @@
+ * 
+ * Requires a {@link EV3ConnectionManager} in the instance to work.
+ * 
+ * Copyright © 2021. TIBCO Software Inc. This file is subject to the license terms contained in the license file that is
```
This is usually at the top before `package`.

kjsmita6 commented on components/mindstorms-ev3/src/main/java/com/example/ev3/EV3SharedObject.java:
```diff
@@ -0,0 +1,417 @@
+        // let's use that.
+        //
+        if (possibleMatches.size() == 1) {
+            EV3SharedObject obj = possibleMatches.get(0);
```
Leftover from debugging?

kjsmita6 commented on components/mindstorms-ev3/src/main/java/com/example/ev3/EV3StatusAdapter.java:
```diff
@@ -0,0 +1,765 @@
+    private boolean StreamPort3;
+    private boolean StreamPort4;
+
+    private int StreamPortAmode = 0;
```
Looks like these are also unused, along with displayName below.

kjsmita6 commented on components/mindstorms-ev3/src/main/java/com/example/ev3/EV3StatusAdapter.java:
```diff
@@ -0,0 +1,765 @@
+        super();
+        setPortHints(inputPorts, outputPorts);
+
+        setDisplayName("MINDSTORMS Status Adapter for EV3");
```
Could replace this with your displayName variable. If so, make it a final variable.

kjsmita6 commented on components/mindstorms-ev3/src/main/java/com/example/ev3/EV3SharedObject.java:
```diff
@@ -0,0 +1,417 @@
+//
+// Copyright (c) 2004-2009 StreamBase Systems, Inc. All rights reserved.
```
StreamBase Systems is now TIBCO :)

TIBCOjstinton replied:
Good catch : )  Should look like this now:

//
// Copyright (c) 2010-2021 TIBCO Software Inc. All rights reserved.
//

kjsmita6 commented on components/mindstorms-ev3/src/main/java/com/example/ev3/RobotPort.java:
```diff
@@ -0,0 +1,75 @@
+    private byte address;
+    private Schema schema;
+
+    public RobotPort(String name, String mode, boolean streaming, byte address, Schema schema, SensorTypeEnum sensor) {
```
Try to keep the overloaded constructor's parameters in the same order. For example: `name`, `streaming`, `address`, and `schema` are part of each constructor so they should be the first 4 arguments. Anything extra should go after (in the same order if they are repeated, like `mode`.

kjsmita6 commented on components/mindstorms-ev3/src/site/markdown/index.md:
```diff
@@ -0,0 +1,3 @@
+# EV3
```
Should this be filled in with something?

oblaufus replied:
I don't actually know. Is this supposed to be the same as the readme?

kjsmita6 replied:
On some other community components it looks like it is, but on others, it isn't. So I'll say yes (@TIBCOjstinton ?)

TIBCOjstinton replied:
I don't think there is a good convention for this yet.  We use the README.md in the root of the project for all of our links and the sample importer to load by default. The /src/site/markdown folder and files are generated by default by our maven archetype.  I think leaving it with default comments, or simply deleting the /src/site folder in general are fine.  Just make sure it hasn't been copy/pasted from another project and therefore has incorrect data in it.

kjsmita6 commented on components/mindstorms-ev3/src/site/site.xml:
```diff
@@ -0,0 +1,30 @@
+    </custom>
+
+    <body>
+        <menu name="Overview">
```
Not sure if these work with markdown files. If not doesn't really matter; the files are there anyways that's what matters.

kjsmita6 commented:
A few more things:
- Missing license ("LICENSE") file, see other repos for example. Should be BSD-3 but a bunch of other licenses are a TIBCO one (see https://github.com/TIBCOSoftware/tibco-streaming-community/pull/15#issuecomment-891826298) @TIBCOjstinton 
- Update components/pom.xml and add your adapter to the components
- Update components/README.md with a link to your readme

kjsmita6 commented:
@TIBCOjstinton I think this is all set to merge, can you take one final look to see if I missed anything? Only lingering problem is that the adapters have no icons and are using defaults.

oblaufus commented:
I can add icons, though I might have questions for you about that. Also there's actually one last feature I'd like to add (or more of, an already existing feature that isn't implemented in one last place I think it should be)

TIBCOjstinton commented:
@oblaufus - There are a few small legal issues. I think.
1) LICENSE does not exactly match the BSD license found here: https://confluence.tibco.com/pages/viewpage.action?spaceKey=EP&title=Contributing+to+GitHub+Community+Samples
2) Copyright statements at the top of Java files should be the first line of the file (before the package and other comments) and should match this format:
`/*
* Copyright © 2021. TIBCO Software Inc.
* This file is subject to the license terms contained
* in the license file that is distributed with this file.
*/`

Looking into the actual meat of the project code, I see:
1) The src\test\java\com\example\ev3\TestCase.java doesn't work, so we can just delete that probably. No requirement to ship it.
2) When I try to start the Demo.sbapp (which I expect won't work because I don't have a Mindstorm!) I see the following error that seems more indictive of a project/library configuration issue:
` Native Library bluecove_x64 not available
 javax.bluetooth.BluetoothStateException: BlueCove library bluecove not available;
 resource not found bluecove_x64.dll
 load [bluecove_x64] no bluecove_x64 in java.library.path: [C:/workspaces/10_6_1_Community_Samples/.nodes/mindstorms-ev3_Demo.jstinton/deploy, C:/workspaces/10_6_1_Community_Samples/.nodes/mindstorms-ev3_Demo.jstinton/devlib, C:/workspaces/10_6_1_Community_Samples/.nodes/mindstorms-ev3_Demo.jstinton/prodlib, C:/TIBCO/sb-cep/10.6/distrib/kabira/lib, C:/TIBCO/sb-cep/10.6/jdk/bin/server, C:/TIBCO/sb-cep/10.6/jdk/bin, C:/cygwin64/bin, C:/opt/jnd/tools/hudson.model.JDK/jdk11.0.8/lib, C:/opt/jnd/tools/hudson.model.JDK/jdk11.0.8/bin/server, C:/TIBCO/sb-cep/10.6/distrib/tibco/be-x/lib, C:/TIBCO/sb-cep/10.6/3rdparty/windows/msvc/12.0_x64, C:/TIBCO/sb-cep/10.6/3rdparty/windows/openssl/1.1.1d_x64/lib, C:/TIBCO/sb-cep/10.6/3rdparty/windows/openssl/1.1.1d_x64/bin, C:/TIBCO/sb-cep/10.6/3rdparty/windows/pccts/1.33_x64/lib, C:/TIBCO/sb-cep/10.6/3rdparty/windows/xerces-c/3.1.1_x64/bin, C:/TIBCO/sb-cep/10.6/3rdparty/windows/msvc/10.0_x64, C:/TIBCO/sb-cep/10.6/bin64, C:/Program Files (x86)/Microsoft SDKs/Azure/CLI2/wbin, C:/Python27/, C:/Python27/Scripts, C:/WINDOWS/system32, C:/WINDOWS, C:/WINDOWS/System32/Wbem, C:/WINDOWS/System32/WindowsPowerShell/v1.0/, C:/Program Files/ThinkPad/Bluetooth Software/, C:/Program Files/ThinkPad/Bluetooth Software/syswow64, C:/Program Files/Broadcom/WHL/, C:/Program Files/Broadcom/WHL/syswow64, C:/Program Files/Broadcom/WHL/SysWow64/, C:/Program Files/Broadcom/WHL/SysWow64/syswow64, C:/Program Files (x86)/Common Files/Lenovo, C:/Program Files (x86)/Common Files/Ulead Systems/MPEG, C:/SWTOOLS/ReadyApps, C:/Program Files (x86)/Intel/Services/IPT/, C:/Program Files/Intel/WiFi/bin/, C:/Program Files/Common Files/Intel/WirelessCommon/, C:/Program Files (x86)/Lenovo/Access Connections/, C:/Program Files/TortoiseSVN/bin, C:/cygwin/bin, C:/TIBCO/as/4.4/bin/, C:/TIBCO/ftl/6.4/bin/, C:/Program Files/PuTTY/, C:/Program Files/MIT/Kerberos/bin, C:/WINDOWS/System32/OpenSSH/, C:/nodejs/, C:/ProgramData/chocolatey/bin, C:/Program Files/Java/jdk1.8.0_201/bin, C:/apache-ant-1.10.7/bin, C:/HashiCorp/Vagrant/bin, C:/Program Files/TortoiseGit/bin, C:/Program Files/Amazon/AWSCLI/bin/, C:/AWS, C:/Program Files/Docker/Docker/resources/bin, C:/ProgramData/DockerDesktop/version-bin, C:/Users/jstinton/AppData/Local/Microsoft/WindowsApps, C:/Users/jstinton/AppData/Roaming/npm, C:/Users/jstinton/AppData/Local/Programs/Microsoft VS Code/bin, C:/Users/jstinton/AppData/Local/GitHubDesktop/bin, C:\workspaces\10_6_1_Community_Samples\.nodes\mindstorms-ev3_Demo.jstinton\prodlib]
 	at com.intel.bluetooth.BlueCoveImpl.loadNativeLibraries(BlueCoveImpl.java:403)
 	at com.intel.bluetooth.BlueCoveImpl.setBluetoothStack(BlueCoveImpl.java:1004)
 	at com.intel.bluetooth.BlueCoveImpl.detectStack(BlueCoveImpl.java:518)
 	at com.intel.bluetooth.BlueCoveImpl.access$500(BlueCoveImpl.java:70)
 	at com.intel.bluetooth.BlueCoveImpl$1.run(BlueCoveImpl.java:1085)
 	at java.base/java.security.AccessController.doPrivileged(Native Method)
 	at com.intel.bluetooth.BlueCoveImpl.detectStackPrivileged(BlueCoveImpl.java:1083)
 	at com.intel.bluetooth.BlueCoveImpl.getBluetoothStack(BlueCoveImpl.java:1076)
 	at com.intel.bluetooth.MicroeditionConnector.openImpl(MicroeditionConnector.java:189)
 	at com.intel.bluetooth.MicroeditionConnector.open(MicroeditionConnector.java:156)
 	at javax.microedition.io.Connector.open(Connector.java:83)
 	at com.j4ev3.desktop.BluetoothComm.connect(BluetoothComm.java:28)
 	at com.j4ev3.desktop.BluetoothComm.<init>(BluetoothComm.java:21)
 	at com.tibco.ep.community.components.ev3.EV3SharedObject.run(EV3SharedObject.java:238)
 	at com.tibco.ep.community.components.ev3.EV3ConnectionManager.init(EV3ConnectionManager.java:94)
 	at com.streambase.sb.runtime.OperatorCaretakerImpl.lambda$init$2(OperatorCaretakerImpl.java:360)
 	at com.streambase.sb.codegen.artifact.ArtifactTaskManager.runTasks(ArtifactTaskManager.java:272)
 	at com.streambase.sb.runtime.MainModule.initMainImpl(MainModule.java:386)
 	at com.streambase.sb.runtime.MainModule.initMain(MainModule.java:342)
 	at com.streambase.sb.containers.Container$InstantiateMod.executeTaskImpl(Container.java:617)
 	at com.streambase.sb.client.internal.AbstractTransaction.run(AbstractTransaction.java:48)
 	at com.kabira.platform.Transaction.execute(Transaction.java:476)
 	at com.streambase.sb.client.internal.AbstractTransaction.start(AbstractTransaction.java:144)
 	at com.streambase.sb.client.internal.AbstractTransaction.executeTask(AbstractTransaction.java:75)
 	at com.streambase.sb.containers.Container.doInstantiateModule(Container.java:576)
 	at com.streambase.sb.containers.Container.init(Container.java:565)
 	at com.streambase.sb.containers.ContainerManager.addContainerImpl(ContainerManager.java:327)
 	at com.streambase.sb.containers.ContainerManager.addContainer(ContainerManager.java:294)
 	at com.streambase.sb.sbd.StreamBaseNode.addContainer(StreamBaseNode.java:497)
 	at com.streambase.sb.sbd.StreamBaseServer.addAllContainers(StreamBaseServer.java:601)
 	at com.streambase.sb.sbd.StreamBaseServer.loadDeploy(StreamBaseServer.java:629)
 	at com.streambase.sb.sbd.StreamBaseServer.setupStartupApps(StreamBaseServer.java:693)
 	at com.streambase.sb.sbd.StreamBaseServer.loadAppsOnStartup(StreamBaseServer.java:758)
 	at com.streambase.sb.sbd.StreamBaseServer.<init>(StreamBaseServer.java:311)
 	at com.tibco.ep.sb.rt.launcher.StreamBaseServerFragment.runFragment(StreamBaseServerFragment.java:158)
 	at com.tibco.ep.sb.rt.launcher.Fragment.run(Fragment.java:80)
`
Is there maybe an extra step besides adding that .jar file to my local maven repo? Do I need to copy out a DLL into my project or system path somewhere?

3) Also during the shutdown I get an NPE. This one looks like it could be easily avoided by first checking for a connection before trying to stop the motors.
`ERROR com.tibco.ep.community.components.ev3.EV3CommandAdapter.default.Java: Error
java.lang.NullPointerException: null
at com.tibco.ep.community.components.ev3.EV3CommandAdapter.shutdown(EV3CommandAdapter.java:405)
at com.streambase.sb.operator.Operator.shutdownImpl(Operator.java:1364)
at com.streambase.sb.runtime.OperatorCaretakerImpl.shutdown(OperatorCaretakerImpl.java:846)
at sb._default_.com.tibco.ep.community.components.ev3.Demo.shutdownImpl(Unknown Source)
at com.streambase.sb.runtime.MainModule.initMainImpl(MainModule.java:394)
at com.streambase.sb.runtime.MainModule.initMain(MainModule.java:342)
at com.streambase.sb.containers.Container$InstantiateMod.executeTaskImpl(Container.java:617)
at com.streambase.sb.client.internal.AbstractTransaction.run(AbstractTransaction.java:48)
at com.kabira.platform.Transaction.execute(Transaction.java:476)
at com.streambase.sb.client.internal.AbstractTransaction.start(AbstractTransaction.java:144)
at com.streambase.sb.client.internal.AbstractTransaction.executeTask(AbstractTransaction.java:75)
at com.streambase.sb.containers.Container.doInstantiateModule(Container.java:576)
at com.streambase.sb.containers.Container.init(Container.java:565)
at com.streambase.sb.containers.ContainerManager.addContainerImpl(ContainerManager.java:327)
at com.streambase.sb.containers.ContainerManager.addContainer(ContainerManager.java:294)
at com.streambase.sb.sbd.StreamBaseNode.addContainer(StreamBaseNode.java:497)
at com.streambase.sb.sbd.StreamBaseServer.addAllContainers(StreamBaseServer.java:601)
at com.streambase.sb.sbd.StreamBaseServer.loadDeploy(StreamBaseServer.java:629)
at com.streambase.sb.sbd.StreamBaseServer.setupStartupApps(StreamBaseServer.java:693)
at com.streambase.sb.sbd.StreamBaseServer.loadAppsOnStartup(StreamBaseServer.java:758)
at com.streambase.sb.sbd.StreamBaseServer.<init>(StreamBaseServer.java:311)
at com.tibco.ep.sb.rt.launcher.StreamBaseServerFragment.runFragment(StreamBaseServerFragment.java:158)
at com.tibco.ep.sb.rt.launcher.Fragment.run(Fragment.java:80)`

oblaufus commented:
1. fixed;
2. fixed.
With regards to the Bluecove issue: That should be included in the J4EV3 jar if you downloaded the "with dependencies" one. May I see what you have under Maven Dependencies > J4EV3, at the bottom? The file "bluecove.dll" that it's looking for should be right in there.
2. Also fixed this: it shouldn't attempt to stop the motors if it's null.

TIBCOjstinton commented:
We confirmed that the library issue (at least on my machine) was due to my BlueTooth adapter being disabled. When I re-enabled it the problem went away.
